tfjs-backend-webgpu/src/backend_webgpu.ts

```diff
@@ -483,18 +483,134 @@ export class WebGPUBackend extends KernelBackend {
+          uniformDataView.setFloat32(
+              dataViewIndex * BYTES_PER_ELEMENT, arrayData, true);
+          dataViewIndex++;
+        }
+      }
+    });
+
+    return uniformDataView;
+  }
+
+  private computePadding(uniformsWithType:
+                             Array<{type: string; data: number[];}>): DataView {
+    let currentOffset = 0;
+    let padding = 0;
+    let dataViewIndex = 0;
+    const dimUniformsData: Array<{type: string; data: number[];}> = [];
+    uniformsWithType.forEach((d, i) => {
+      if (d.data.length === 0) {
+        d.data = [1];
+      }
+      // Complete std140 layout rules are documented here:
+      // tslint:disable-next-line:max-line-length
+      // https://www.khronos.org/registry/OpenGL/specs/gl/glspec45.core.pdf#page=159
+      let baseAlignment: number;
+      switch (d.data.length) {
+        case 0:
+          baseAlignment = 1;
+          break;
+        case 1:
+          baseAlignment = 1;
+          break;
+        case 2:
+          baseAlignment = 2;
+          break;
+        case 3:
+          baseAlignment = 4;
+          break;
+        case 4:
+          baseAlignment = 4;
+          break;
+        default:
+          util.assert(false, () => `Unsupported ${d.data.length}D shape`);
+      }
+
+      padding = Math.ceil(currentOffset / baseAlignment) * baseAlignment -
+          currentOffset;
+      for (let p = 0; p < padding; ++p) {
+        dimUniformsData.push({type: 'int32', data: [0]});
+        dataViewIndex++;
+      }
+      dimUniformsData.push({type: d.type, data: d.data});
+      dataViewIndex = dataViewIndex + d.data.length;
+      currentOffset += d.data.length + padding;
+    });
+
+    return this.numberOrArrayToDataView(dimUniformsData, dataViewIndex);
+  }
+
   public runWebGPUProgram(
       program: webgpu_program.WebGPUProgram, inputs: TensorInfo[],
       outputDtype: DataType,
-      programUniforms?: Uint32Array|Int32Array|Float32Array): TensorInfo {
+      programUniforms?: Array<{type: string; data: number[]}>): TensorInfo {
     const output = this.makeTensorInfo(program.outputShape, outputDtype);
 
-    let uniformDataLength;
-    let uniforms: GPUBindingResource;
-    if (program.uniforms) {
-      // TODO: handle padding of program-specific uniforms
-      uniformDataLength = programUniforms.byteLength;
-      uniforms = this.makeUniforms(programUniforms);
+    // There are four kinds of uniforms: shapes, shape strides, program
+    // size, program defined uniforms.
+    const bufferShapes = inputs.concat(output).map(d => d.shape);
+    let uniformsWithType = bufferShapes.map(d => {
+      return {type: 'int32', data: d};
+    });
+    const strides = util.computeStrides(output.shape);
+    uniformsWithType.push({type: 'int32', data: strides});
+    if (program.size != null) {
+      uniformsWithType.push({type: 'int32', data: [program.size]});
+    }
+    if (programUniforms) {
+      uniformsWithType = [...uniformsWithType, ...programUniforms];
```

Comment:
Do we merge these two uniform blocks together here? If yes, why? Their attributes are different.
If we do not merge them together, do we only need to handle `int` type in numberOrArrayToDataView function? And do this uniform block data could be packed into Array<number>, instead of Array<{type: string; data: number[]}>? We do not need to encode and decode it.

Reply:
This will put all uniforms inside one uniform block. These uniforms data are small, so I think no need to create more than one slot.
These uniforms are from two different  sources (automaticall generated: shapes, shape strides, program size; and program defined). For one uniform buffer case, if we want to got correct padding, we need get all uniforms first, then compute the padding.